File: script/python/space/binary_search.py
def binary_search(arr, trgt):
    leftindex = 0
    rightindex = len(arr) - 1
    
    while leftindex <= rightindex:
        midindex = (leftindex + rightindex) // 2
        if trgt == arr[midindex]:
            return midindex
        elif trgt < arr[midindex]:
            rightindex = midindex - 1
        else:
            leftindex = midindex + 1
    
    return -1
    

def quicksort(arr):
    if len(arr) <= 1:
        return arr
    pivot = arr[len(arr) // 2]
    mid = [x for x in arr if x == pivot]
    left = [x for x in arr if x < pivot]
    right = [x for x in arr if x > pivot]
    return quicksort(left) + quicksort(mid) + quicksort(right)
    


def main():
    arr = [2,3,1,4,5,0,9,8,17,16,14,18,10,6]
    arr = quicksort(arr)

    while True:
        try: expect = int(input())
        except: 
            ValueError
            continue
        if expect == -1:
            break
        print(binary_search(arr, expect))


if __name__ == "__main__":
    main()

File: compile_lang/C++/space/binary_search/binary_search.cpp
#include "binary_search.h"

int binary_search(std::vector<int> &arr, int target) {
  int left = 0, right = arr.size() - 1;
  while (left <= right) {
    int mid = (left + right) / 2;
    if (arr[mid] == target)
      return mid;
    if (arr[mid] < target)
      left = mid + 1;
    if (arr[mid] > target)
      right = mid - 1;
  }

  return -1;
}
